sklearn/utils/_heap.pyx

```diff
@@ -0,0 +1,154 @@
+
+    Notes
+    -----
+    Arrays are manipulated via a pointer to there first element and their size
+    as to ease the processing of dynamically allocated buffers.
+    """
+    # TODO: In order to support discrete distance metrics, we need to have a
+    # simultaneous sort which breaks ties on indices when distances are identical.
+    # The best might be using a std::stable_sort and a Comparator which might need
+    # an Array of Structures (AoS) instead of the Structure of Arrays (SoA)
+    # currently used.
+    cdef:
+        ITYPE_t pivot_idx, i, store_idx
+        floating pivot_val
+
+    # in the small-array case, do things efficiently
+    if size <= 1:
+        pass
+    elif size == 2:
+        if values[0] > values[1]:
+            dual_swap(values, indices, 0, 1)
+    elif size == 3:
+        if values[0] > values[1]:
+            dual_swap(values, indices, 0, 1)
+        if values[1] > values[2]:
+            dual_swap(values, indices, 1, 2)
+            if values[0] > values[1]:
+                dual_swap(values, indices, 0, 1)
+    else:
+        # Determine the pivot using the median-of-three rule.
+        # The smallest of the three is moved to the beginning of the array,
+        # the middle (the pivot value) is moved to the end, and the largest
+        # is moved to the pivot index.
+        pivot_idx = size // 2
+        if values[0] > values[size - 1]:
+            dual_swap(values, indices, 0, size - 1)
+        if values[size - 1] > values[pivot_idx]:
+            dual_swap(values, indices, size - 1, pivot_idx)
+            if values[0] > values[size - 1]:
+                dual_swap(values, indices, 0, size - 1)
+        pivot_val = values[size - 1]
+
+        # Partition indices about pivot.  At the end of this operation,
+        # pivot_idx will contain the pivot value, everything to the left
+        # will be smaller, and everything to the right will be larger.
+        store_idx = 0
+        for i in range(size - 1):
+            if values[i] < pivot_val:
+                dual_swap(values, indices, i, store_idx)
+                store_idx += 1
+        dual_swap(values, indices, store_idx, size - 1)
+        pivot_idx = store_idx
+
+        # Recursively sort each side of the pivot
+        if pivot_idx > 1:
+            simultaneous_sort(values, indices, pivot_idx)
+        if pivot_idx + 2 < size:
+            simultaneous_sort(values + pivot_idx + 1,
+                              indices + pivot_idx + 1,
+                              size - pivot_idx - 1)
+    return 0
+
+
+cdef inline int heap_push(
+    floating* values,
+    ITYPE_t* indices,
+    ITYPE_t size,
+    floating val,
+    ITYPE_t val_idx,
+) nogil:
+    """Push a tuple (val, val_idx) onto a fixed-size max-heap.
+
+    The max-heap is represented as a Structure of Arrays where:
+     - values is the array containing the data to construct then heap with
+     - indices is the array containing the indices (meta-data) of each value
+
```

Comment:
Maybe a quick pseudo code example would help:


```
heap_push(
    values=[0.1, 0.4, 1.2],
    indices=[42, 1, 5],
    size=3,
    val=0.2,
    val_idx=4,
)
```
would result in the updated:

```python
values == [0.1, 0.2, 0.4]
indices == [42, 4, 1]
```

Reply:
If `val=0.2`, then `val >= values[0]`, and nothing happens.

From my understanding, I think the max heap has the maximum element in the 0th position and then swapping is done resulting in:

```
values = [1.2, 0.4, 0.1]
indices = [42, 1, 5]
heap_push(
    values=values,
    indices=indices,
    size=3,
    val=0.2,
    val_idx=4,
)
```

results in:

```python
values == [0.4, 0.2, 0.1]
indices == [1, 4, 5]
```